haskell.nix/flake.nix

```diff
@@ -0,0 +1,116 @@
+      url = "github:edolstra/flake-compat";
+      flake = false;
+    };
+    nixpkgs.url = "github:serokell/nixpkgs";
+    haskell-nix.url = "github:serokell/haskell.nix/serokell-latest";
+    hackage-nix = {
+      url = "github:input-output-hk/hackage.nix";
+      flake = false;
+    };
+    stackage-nix = {
+      url = "github:input-output-hk/stackage.nix";
+      flake = false;
+    };
+
+    # if you want to run weeder:
+    ##haskell-nix-weeder = {
+    ##  url = "github:serokell/haskell-nix-weeder";
+    ##  flake = false;
+    ##};
+  };
+
+  outputs = { self, nixpkgs, haskell-nix, hackage-nix, stackage-nix, ... }:
+
+  # for weeder:
+  ##outputs = { self, nixpkgs, haskell-nix, hackage-nix, stackage-nix, haskell-nix-weeder, ... }:
+
+    let
+      haskellNix = import haskell-nix {
+        sourcesOverride = { hackage = hackage-nix; stackage = stackage-nix; };
+      };
+
+      pkgs = import nixpkgs haskellNix.nixpkgsArgs;
+      lib = pkgs.lib;
+
+      # invoke haskell.nix
+      hs-pkgs = pkgs.haskell-nix.stackProject {
+        src = pkgs.haskell-nix.haskellLib.cleanGit {
+          name = "pataq-package";
+          src = ./.;
+        };
+
+        # haskell.nix configuration
+        modules = [{
+          packages.pataq-package = {
+            ghcOptions = [
+              # fail on warnings
+              "-Werror"
+              # disable optimisations, we don't need them if we don't package or deploy the executable
+              "-O0"
+
+              # for weeder: produce *.dump-hi files
+              ##"-ddump-to-file" "-ddump-hi"
+            ];
+
+            # for weeder: collect all *.dump-hi files
+            ##postInstall = weeder-hacks.collect-dump-hi-files;
+          };
+
+        }];
+      };
+
+      hs-pkg = hs-pkgs.pataq-package;
+
+      # returns the list of all components for a package
+      get-package-components = pkg:
+        # library
+        lib.optional (pkg ? library) pkg.library
+        # haddock
+        ++ lib.optional (pkg ? library) pkg.library.haddock
+        # exes, tests and benchmarks
+        ++ lib.attrValues pkg.exes
+        ++ lib.attrValues pkg.tests
+        ++ lib.attrValues pkg.benchmarks;
+
+      # all components for the current haskell package
+      all-components = get-package-components hs-pkg.components;
+
+      # for weeder:
+      ##weeder-hacks = import haskell-nix-weeder { inherit pkgs; };
+
+      # nixpkgs has weeder 2, but we use weeder 1
+      ##weeder-legacy = pkgs.haskellPackages.callHackageDirect {
+      ##  pkg = "weeder";
+      ##  ver = "1.0.9";
+      ##  sha256 = "0gfvhw7n8g2274k74g8gnv1y19alr1yig618capiyaix6i9wnmpa";
+      ##} {};
+
+      # a derivation which generates a script for running weeder
+      ##weeder-script = weeder-hacks.weeder-script {
+      ##  weeder = weeder-legacy;
+      ##  hs-pkgs = hs-pkgs;
+      ##  local-packages = [
+      ##    { name = "pataq-package"; subdirectory = "."; }
+      ##  ];
+      ##};
+
+    in {
```

Comment:
We don't actually output any of the package's components, which is strange

Reply:
CI builds them through `checks`, it's better not to add them as `packages` if we don't intend to use them for anything, because there'll be an overhead in maintaining them. E.g. if we decide to add parameters to a the package definition, we would need to decide which values to use for `packages`, and to check that it still works

Reply:
I see no reason we can't just at least export executable components under `packages` here (and maybe add an `apps` definition too) as an example and it can just be deleted if not needed, otherwise this is an incomplete template

Reply:
This template is mainly to help devs set up simple CI to check that their haskell projects builds and to run tests. The problem with expanding it to packaging is that there are many ways we may want to package the executables, e.g. making dynamically/statically linked binaries, making packages with extra resource files (like migration scripts), making docker images, making executables for different systems. And for many projects we don't need to package anything at all, we just need tests in CI.

So the way we would do packaging depends a lot on the project, we would have to add some explanation for the devs on how to choose a packaging method, and it's hard to do without a specific practical usecase (e.g. deploying to a server or providing downloadable packages for users). And just using a random way to do it will be confusing, I think, because people won't know what it's used for, and what to do with it when they want to make conflicting changes, like adding parameters to the derivations.

If we want to also make templates for devs for packaging/deploying packages from CI, I think it's better to do it separately, because it's an issue of a big scope by itself.

Reply:
There's already assumptions being made for the components which are used in the flake checks, and I think there is an objectively most generic way to export as a package, which I think should at least be included.

This is a template after all, the idea is to provide a good base which can be easily modified to fit each project, so if you don't want to export packages you can delete that, if you want to do it differently then modify it, and if the default generic methodology is appropriate for your project then you don't need to do anything at all.

I also wouldn't be against adding more than a single example, such as to catch the other use cases that you mentioned too, and disagree that having it separately would be a good idea, as it would end up being incomplete (and require developers to manually merge implementations from both templates) or end up containing all of the same code here copy-pasted. I think this is the most appropriate place to include any of those examples.

Reply:
By separately I just mean in a different issue/PR, so that we can get some basic template ready for people sooner. I'm not sure if a separate template would be better for this.

I'm not sure what would a generic package look like, it would be different for different usecases. And I think it would be confusing to devs to provide an example of a package without explaining when this exact implementation is appropriate, and how to use it for actually deploying to a server or delivering to users, so it does not seem useful to add an example of a package without also adding explanation on how to do these tasks as well. Which looks like a considerate amount of work to me, so I'd really like to not do it right now.